### PR TITLE
Start work for OmitIdentityInJava + add java struct tests

### PR DESCRIPTION
**Pull request by oldergod**
The generated code could get some cleaning obviously...

### CODE REVIEW

File: wire-library/wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java [hunk 1]
```diff
@@ -927,15 +927,25 @@ private MethodSpec messageAdapterEncodedSize(NameAllocator nameAllocator, Messag
...
     String leading = "return";
     for (Field field : type.getFieldsAndOneOfFields()) {
       int fieldTag = field.getTag();
       String fieldName = nameAllocator.get(field);
       CodeBlock adapter = adapterFor(field, nameAllocator);
-      result.addCode("$L $L.encodedSizeWithTag($L, ", leading, adapter, fieldTag)
+      result.addCode("$L ", leading);
+      boolean omitIdentity = field.getEncodeMode().equals(Field.EncodeMode.OMIT_IDENTITY);
+      if (omitIdentity) {
+        result.addCode("($T.equals(value.$L, $L) ? 0\n  : ", ClassName.get(java.util.Objects.class),
```
Comment (JakeWharton) [inline wire-library/wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java:940]: nit: import type

File: wire-library/wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java [hunk 2]
```diff
@@ -1791,6 +1806,60 @@ private CodeBlock fieldInitializer(ProtoType type, @Nullable Object value) {
...
+          } else {
+            throw new IllegalArgumentException("Unexpected scalar proto type: " + protoType);
+          }
+        } else if (type instanceof MessageType) {
+          return CodeBlock.of("null");
+        } else if (type instanceof EnumType) {
+          return CodeBlock.of("$T.$L", typeName(protoType),
+              ((EnumType) type).constant(0).getName());
+        }
+      case REQUIRED:
```
Comment (JakeWharton) [inline wire-library/wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java:1857]: throw?

File: wire-library/wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java (hunk 1 above)
Comment (oldergod) [reply wire-library/wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java]: fixed

File: wire-library/wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java (hunk 2 above)
Comment (oldergod) [reply wire-library/wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java:1858]: fixed

File: wire-library/wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java [hunk 3]
```diff
@@ -1791,6 +1807,61 @@ private CodeBlock fieldInitializer(ProtoType type, @Nullable Object value) {
...
+        return CodeBlock.of("null");
+      case OMIT_IDENTITY:
+        ProtoType protoType = field.getType();
+        Type type = schema.getType(protoType);
+        if (protoType.equals(ProtoType.STRUCT_NULL)) {
+          return CodeBlock.of("null");
+        } else if (field.isOneOf()) {
+          return CodeBlock.of("null");
+        } else if (protoType.isScalar()) {
+          if (protoType.equals(ProtoType.BOOL)) {
```
Comment (swankjesse) [inline wire-library/wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java:1827]: should this be a `Map<ProtoType, CodeBlock>` ?

Comment (oldergod) [reply wire-library/wire-java-generator/src/main/java/com/squareup/wire/java/JavaGenerator.java:1827]: Let's do it. Kotlin would need it to, I'll follow up

**Pull request merged**